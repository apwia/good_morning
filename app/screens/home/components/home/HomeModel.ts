import React from "react";
import Alarm from "../../../../model/Alarm";

type HomeModel = {
    alarms: Alarm[],
    addAction: any,
    showTimePick: boolean,
    setShowTimePick: React.Dispatch<React.SetStateAction<boolean>>,
    newAlarm: Alarm,
    setNewAlarm: React.Dispatch<React.SetStateAction<Alarm>>,
    openTimePickDialog: any,
    setShouldRefreshAlarms: React.Dispatch<React.SetStateAction<boolean>>,
}

export default HomeModel;