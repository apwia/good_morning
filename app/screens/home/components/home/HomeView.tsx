import React from 'react';

import { FlatList, Image, SafeAreaView, Text, TouchableOpacity, View } from 'react-native';

import useHomeViewModel from './HomeViewModel';
import Alarm from '../../../../model/Alarm';
import Styles from './HomeStyles';
import AlarmView from '../alarm/AlarmView';
import SeparatorView from '../separator/SeparatorView';
import TimePickerView from '../time_picker/TimePickerView';

const HomeView = ( {navigation}: any) => {

  const {alarms, addAction, showTimePick, setShowTimePick, newAlarm, openTimePickDialog, setShouldRefreshAlarms} = 
    useHomeViewModel();

  const RenderAlarm = (renderProps: Alarm) => (
    <AlarmView 
      key={renderProps.id}
      alarm={renderProps}
      setShouldRefreshAlarms={setShouldRefreshAlarms}/>
  );
  
  return (

    <SafeAreaView style={Styles.background}>
    
    <View style={Styles.titleView}>
      <Text style={Styles.title}>Alarm Clock</Text>

      <TouchableOpacity onPress={()=> navigation.navigate('About')} testID='navigateToAbout'>
        <Image source={require("../../../../assets/images/AboutIcon.png")} style={Styles.aboutImage} />
      </TouchableOpacity>

    </View>


      <FlatList
        data={alarms}
        renderItem={({ item }) => <RenderAlarm {...item}/> }
        keyExtractor={alarm => alarm.id+""}
        ItemSeparatorComponent={() => <SeparatorView />}/>

      <View style={Styles.addButtonView}>

        <TouchableOpacity onPress={openTimePickDialog}>
          <Image style={Styles.addButton} source={require("../../../../assets/images/Add.png")}/>
        </TouchableOpacity>

        {showTimePick && 
          <TimePickerView
            alarm={newAlarm} 
            action={addAction} 
            whenFinish={() => {setShowTimePick(false)}} />}
        
      </View>

    </SafeAreaView>
);
}

export default HomeView;