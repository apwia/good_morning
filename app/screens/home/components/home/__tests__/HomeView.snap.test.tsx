import 'react-native';
import React from 'react';

import renderer from 'react-test-renderer';
import HomeView from '../HomeView';
import HomeModel from '../HomeModel';

jest.mock('../HomeViewModel', () => {
  const fnHomeViewModel = jest.fn();
  fnHomeViewModel.mockImplementation(
    () => ({
      alarms: [{ id: 0, time: new Date(123456788), active: false, periodicity: 'every day' },
               { id: 1, time: new Date(123456799), active: true, periodicity: 'every day' }],
      addAction: jest.fn(),
      showTimePick: true,
      setShowTimePick: jest.fn(),
      newAlarm: { id: 0, time: new Date(123456789), active: false, periodicity: 'every day' },
      setNewAlarm: jest.fn(),
      openTimePickDialog: jest.fn(),
      setShouldRefreshAlarms: jest.fn()
    } as HomeModel),
  );

  return fnHomeViewModel;

});

import useHomeViewModel from '../HomeViewModel';
import { fireEvent, render, screen } from '@testing-library/react-native';

it('shoud render correctly', () => {

  const props = {
    navigation: jest.fn()
  };

  const {toJSON} = renderer.create(<HomeView {...props} />);
  expect(toJSON()).toMatchSnapshot();

});

it('should navigate to About page', () => {

  const navigateFn = jest.fn();

  const navigationMock = {navigation: {navigate: navigateFn}};

  render(<HomeView {...navigationMock}/>)

  fireEvent.press(screen.getByTestId('navigateToAbout'));

  expect(navigateFn).toBeCalledWith('About');

});