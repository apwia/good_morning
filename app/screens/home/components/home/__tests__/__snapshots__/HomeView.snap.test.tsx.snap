// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`shoud render correctly 1`] = `
<RCTSafeAreaView
  style={
    {
      "flex": 1,
      "padding": 5,
    }
  }
>
  <View
    style={
      {
        "flexDirection": "row",
        "padding": 15,
      }
    }
  >
    <Text
      style={
        {
          "color": "#717171",
          "flexGrow": 1,
          "fontSize": 25,
          "textAlign": "center",
        }
      }
    >
      Alarm Clock
    </Text>
    <View
      accessibilityState={
        {
          "busy": undefined,
          "checked": undefined,
          "disabled": undefined,
          "expanded": undefined,
          "selected": undefined,
        }
      }
      accessibilityValue={
        {
          "max": undefined,
          "min": undefined,
          "now": undefined,
          "text": undefined,
        }
      }
      accessible={true}
      collapsable={false}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        {
          "opacity": 1,
        }
      }
      testID="navigateToAbout"
    >
      <Image
        source={
          {
            "testUri": "../../../app/assets/images/AboutIcon.png",
          }
        }
        style={
          {
            "alignItems": "center",
            "flexDirection": "row-reverse",
            "height": 28,
            "padding": 10,
            "width": 28,
          }
        }
      />
    </View>
  </View>
  <RCTScrollView
    ItemSeparatorComponent={[Function]}
    data={
      [
        {
          "active": false,
          "id": 0,
          "periodicity": "every day",
          "time": 1970-01-02T10:17:36.788Z,
        },
        {
          "active": true,
          "id": 1,
          "periodicity": "every day",
          "time": 1970-01-02T10:17:36.799Z,
        },
      ]
    }
    getItem={[Function]}
    getItemCount={[Function]}
    keyExtractor={[Function]}
    onContentSizeChange={[Function]}
    onLayout={[Function]}
    onMomentumScrollBegin={[Function]}
    onMomentumScrollEnd={[Function]}
    onScroll={[Function]}
    onScrollBeginDrag={[Function]}
    onScrollEndDrag={[Function]}
    removeClippedSubviews={false}
    renderItem={[Function]}
    scrollEventThrottle={50}
    stickyHeaderIndices={[]}
    viewabilityConfigCallbackPairs={[]}
  >
    <View>
      <View
        onFocusCapture={[Function]}
        onLayout={[Function]}
        style={null}
      >
        <View>
          <View
            collapsable={false}
            forwardedRef={[Function]}
            handlerTag={1}
            handlerType="PanGestureHandler"
            onGestureHandlerEvent={[Function]}
            onGestureHandlerStateChange={[Function]}
            onLayout={[Function]}
            style={
              {
                "flex": 1,
                "overflow": "hidden",
              }
            }
          >
            <View
              collapsable={false}
              style={
                {
                  "bottom": 0,
                  "flexDirection": "row",
                  "left": 0,
                  "position": "absolute",
                  "right": 0,
                  "top": 0,
                  "transform": [
                    {
                      "translateX": -10000,
                    },
                  ],
                }
              }
            >
              <View
                collapsable={false}
                style={
                  {
                    "alignItems": "center",
                    "flexDirection": "row-reverse",
                    "transform": [
                      {
                        "scale": 0,
                      },
                    ],
                  }
                }
              >
                <View
                  accessibilityState={
                    {
                      "busy": undefined,
                      "checked": undefined,
                      "disabled": undefined,
                      "expanded": undefined,
                      "selected": undefined,
                    }
                  }
                  accessibilityValue={
                    {
                      "max": undefined,
                      "min": undefined,
                      "now": undefined,
                      "text": undefined,
                    }
                  }
                  accessible={true}
                  collapsable={false}
                  focusable={true}
                  onClick={[Function]}
                  onResponderGrant={[Function]}
                  onResponderMove={[Function]}
                  onResponderRelease={[Function]}
                  onResponderTerminate={[Function]}
                  onResponderTerminationRequest={[Function]}
                  onStartShouldSetResponder={[Function]}
                  style={
                    {
                      "opacity": 1,
                    }
                  }
                >
                  <Image
                    source={
                      {
                        "testUri": "../../../app/assets/images/TrashBin.png",
                      }
                    }
                    style={
                      {
                        "height": 50,
                        "width": 50,
                      }
                    }
                  />
                </View>
              </View>
              <View
                onLayout={[Function]}
              />
            </View>
            <View
              collapsable={false}
              forwardedRef={[Function]}
              handlerTag={2}
              handlerType="TapGestureHandler"
              onGestureHandlerEvent={[Function]}
              onGestureHandlerStateChange={[Function]}
              pointerEvents="auto"
              style={
                {
                  "transform": [
                    {
                      "translateX": -0,
                    },
                  ],
                }
              }
            >
              <View
                style={
                  {
                    "flexDirection": "row",
                  }
                }
              >
                <View
                  style={
                    {
                      "flexGrow": 1,
                    }
                  }
                >
                  <Text
                    onPress={[Function]}
                    style={
                      {
                        "color": "#7f7f7f",
                        "fontSize": 35,
                      }
                    }
                  >
                    07:17
                  </Text>
                  <Text
                    style={
                      {
                        "color": "#a4a4a4",
                      }
                    }
                  >
                    every day
                  </Text>
                </View>
                <View
                  style={
                    {
                      "alignItems": "center",
                      "flexDirection": "row-reverse",
                      "padding": 10,
                    }
                  }
                >
                  <RCTSwitch
                    accessibilityRole="switch"
                    onChange={[Function]}
                    onResponderTerminationRequest={[Function]}
                    onStartShouldSetResponder={[Function]}
                    style={
                      {
                        "height": 31,
                        "width": 51,
                      }
                    }
                    testID="switch"
                    value={false}
                  />
                </View>
              </View>
            </View>
          </View>
        </View>
        <View
          style={
            {
              "borderBottomColor": "#737373",
              "borderBottomWidth": 0.5,
              "marginVertical": 8,
            }
          }
        />
      </View>
      <View
        onFocusCapture={[Function]}
        onLayout={[Function]}
        style={null}
      >
        <View>
          <View
            collapsable={false}
            forwardedRef={[Function]}
            handlerTag={3}
            handlerType="PanGestureHandler"
            onGestureHandlerEvent={[Function]}
            onGestureHandlerStateChange={[Function]}
            onLayout={[Function]}
            style={
              {
                "flex": 1,
                "overflow": "hidden",
              }
            }
          >
            <View
              collapsable={false}
              style={
                {
                  "bottom": 0,
                  "flexDirection": "row",
                  "left": 0,
                  "position": "absolute",
                  "right": 0,
                  "top": 0,
                  "transform": [
                    {
                      "translateX": -10000,
                    },
                  ],
                }
              }
            >
              <View
                collapsable={false}
                style={
                  {
                    "alignItems": "center",
                    "flexDirection": "row-reverse",
                    "transform": [
                      {
                        "scale": 0,
                      },
                    ],
                  }
                }
              >
                <View
                  accessibilityState={
                    {
                      "busy": undefined,
                      "checked": undefined,
                      "disabled": undefined,
                      "expanded": undefined,
                      "selected": undefined,
                    }
                  }
                  accessibilityValue={
                    {
                      "max": undefined,
                      "min": undefined,
                      "now": undefined,
                      "text": undefined,
                    }
                  }
                  accessible={true}
                  collapsable={false}
                  focusable={true}
                  onClick={[Function]}
                  onResponderGrant={[Function]}
                  onResponderMove={[Function]}
                  onResponderRelease={[Function]}
                  onResponderTerminate={[Function]}
                  onResponderTerminationRequest={[Function]}
                  onStartShouldSetResponder={[Function]}
                  style={
                    {
                      "opacity": 1,
                    }
                  }
                >
                  <Image
                    source={
                      {
                        "testUri": "../../../app/assets/images/TrashBin.png",
                      }
                    }
                    style={
                      {
                        "height": 50,
                        "width": 50,
                      }
                    }
                  />
                </View>
              </View>
              <View
                onLayout={[Function]}
              />
            </View>
            <View
              collapsable={false}
              forwardedRef={[Function]}
              handlerTag={4}
              handlerType="TapGestureHandler"
              onGestureHandlerEvent={[Function]}
              onGestureHandlerStateChange={[Function]}
              pointerEvents="auto"
              style={
                {
                  "transform": [
                    {
                      "translateX": -0,
                    },
                  ],
                }
              }
            >
              <View
                style={
                  {
                    "flexDirection": "row",
                  }
                }
              >
                <View
                  style={
                    {
                      "flexGrow": 1,
                    }
                  }
                >
                  <Text
                    onPress={[Function]}
                    style={
                      {
                        "color": "#7f7f7f",
                        "fontSize": 35,
                      }
                    }
                  >
                    07:17
                  </Text>
                  <Text
                    style={
                      {
                        "color": "#a4a4a4",
                      }
                    }
                  >
                    every day
                  </Text>
                </View>
                <View
                  style={
                    {
                      "alignItems": "center",
                      "flexDirection": "row-reverse",
                      "padding": 10,
                    }
                  }
                >
                  <RCTSwitch
                    accessibilityRole="switch"
                    onChange={[Function]}
                    onResponderTerminationRequest={[Function]}
                    onStartShouldSetResponder={[Function]}
                    style={
                      {
                        "height": 31,
                        "width": 51,
                      }
                    }
                    testID="switch"
                    value={true}
                  />
                </View>
              </View>
            </View>
          </View>
        </View>
      </View>
    </View>
  </RCTScrollView>
  <View
    style={
      {
        "alignItems": "center",
      }
    }
  >
    <View
      accessibilityState={
        {
          "busy": undefined,
          "checked": undefined,
          "disabled": undefined,
          "expanded": undefined,
          "selected": undefined,
        }
      }
      accessibilityValue={
        {
          "max": undefined,
          "min": undefined,
          "now": undefined,
          "text": undefined,
        }
      }
      accessible={true}
      collapsable={false}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        {
          "opacity": 1,
        }
      }
    >
      <Image
        source={
          {
            "testUri": "../../../app/assets/images/Add.png",
          }
        }
        style={
          {
            "height": 65,
            "width": 65,
          }
        }
      />
    </View>
    <RNDateTimePicker
      date={123456789}
      displayIOS="default"
      enabled={true}
      is24Hour={true}
      mode="time"
      onChange={[Function]}
      onPickerDismiss={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
    />
  </View>
</RCTSafeAreaView>
`;
